### PR TITLE
add python package handler

### PR DESCRIPTION
**Pull request by g-arjones**

### CODE REVIEW

File: lib/autobuild/packages/python.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    def self.python(opts, &proc)
+        Python.new(opts, &proc)
+    end
+
+    # Handler class to build python-based packages
+    class Python < Configurable
+        attr_accessor   :buildflags
+        attr_accessor   :installflags
+
+        @builddir = 'build'
```
Comment (doudou) [inline lib/autobuild/packages/python.rb]: This does ... nothing

File: lib/autobuild/packages/python.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        def prepare_for_forced_build
+            super
+            FileUtils.rm_f configurestamp
+            @forced = true
+        end
+
+        # There is nothing to configure
+        def configure
+            super {}
```
Comment (doudou) [inline lib/autobuild/packages/python.rb]: `super <empty block>` ? Leftover ?

File: lib/autobuild/packages/python.rb (hunk 1 above)
Comment (g-arjones) [reply lib/autobuild/packages/python.rb]: Yeah.. I was wondering why you did that on both cmake and autotools package handlers. Will remove :+1:

File: lib/autobuild/packages/python.rb (hunk 2 above)
Comment (g-arjones) [reply lib/autobuild/packages/python.rb]: The base class yields without checking block_given?.

**Comment by g-arjones** [comment]
> If you don't mind, I would suggest providing a non-install mode.

Absolutely, no problem.

> Moreover, shouldn't the package actually set up some env vars for the install folder ?

For the no-install mode, definitely. Other than that, my understanding is that autoproj already adds `#{prefix}/lib/python2.7/site-packages` to `PYTHONPATH`. Am I wrong?

**Comment by doudou** [comment]
> Am I wrong?

Seems so, yes. The method that auto-updates these kind of envvars is [`Autobuild::Environment#add_prefix`](https://github.com/rock-core/autobuild/blob/master/lib/autobuild/environment.rb#L674) and it's not handled there. Probably better to update this method, though.

The next problem will be to decide how to handle the python version. 2.x is almost end of life, but still very used.

**Comment by g-arjones** [comment]
> The next problem will be to decide how to handle the python version. 2.x is almost end of life, but still very used.

Check for python2 and python3 in $PATH and do a `python[2-3] -V` to add both versions to `PYTHONPATH`?

**Comment by doudou** [comment]
Here, they've been using `#{name}`

**Comment by doudou** [comment]
> add both versions to PYTHONPATH?

Does this work ? I would have assumed that you must set up PYTHONPATH to the "right" path. Not sure if there aren't python libraries that have different contents between the two.

I really don't know.

**Comment by g-arjones** [comment]
> Not sure if there aren't python libraries that have different contents between the two

What python3 modules that are compatible with python2 usually do is add a module.pth lib/python2.7/site-packages pointing to the python3 path, they don't add different files to both paths.

But now that you brought this up, the problem I see with having both paths in PYTHONPATH is that python2 might end up importing incompatible python3 stuff...

**Comment by g-arjones** [comment]
OK, after doing some research, it seems that there is no good way to handle multiple python versions in the same workspace. For that, developers often use tools such as [virtualenv](https://virtualenv.pypa.io/en/stable/). I am not sure if autoproj should support that.

What we should do instead is set `PYTHONUSERBASE` (rather then `PYTHONPATH`) to prefix, and let python itself decide where it will look for modules. This is actually what tools like setuptools and pip do (you set the prefix and the installation path is dependent on the version of the interpreter being used).

For non-install packages, we set `PYTHONPATH` and let the user handle different python versions if necessary.

In the future, we might provide a config option to let the user set python's executable path to allow the user to use a python version different from the system's "default".

Thoughts?

**Comment by doudou** [comment]
> Thoughts?

Sounds like a plan !

In the longer term, I wanted to try and use pipenv as a Python equivalent to Bundler to install the pip packages. For user-provided packages, though, I think that the plan you laid out makes perfect sense.

**Comment by g-arjones** [comment]
> Sounds like a plan !

Unfortunately, `PipManager` [is hijacking](https://github.com/rock-core/autoproj/blob/181a7142c283fb811de1b11705072d5e0e48f2d2/lib/autoproj/package_managers/pip_manager.rb#L8) `PYTHONUSERBASE`. Is there are reason to have a separate prefix for pip alone? Usually, python modules are all in the same prefix regardless of the tool used to install them (setuptools, easy_install, pip...)

**Comment by doudou** [comment]
> Usually, python modules are all in the same prefix regardless of the tool used to install them (setuptools, easy_install, pip...)

Well ... actually not under autoproj. `config.separate_prefixes` requires the packages to be built and installed in separate install folders (or not installed at all), and supporting it is required for the Jenkins plugin (it makes it trivial to propagate the build artifacts from job to job).

It looks like python itself could help with building a valid PYTHONPATH. There's [`site.addsitedir`](https://docs.python.org/3/library/site.html#site.addsitedir). 

Otherwise, the simplest at this stage would be to not install the package at all.

**Comment by g-arjones** [comment]
> It looks like python itself could help with building a valid PYTHONPATH

I will give that a try.

> Otherwise, the simplest at this stage would be to not install the package at all.

We can do that if nothing else works.. Some python modules require building and adding a builddir to PYTHONPATH is kinda of an anti-pattern within autoproj...

**Comment by doudou** [comment]
Actually, `python3 -m site --user-site` would allow to infer the relative path from the prefix to the entry in PYTHONPATH. Better than `add_site` since it could be called once.

**Comment by g-arjones** [comment]
That's the plan.. :+1:

**Comment by g-arjones** [comment]
Updated. I have also added a couple of commits to address the issues you mentioned in the beginning of the thread

**Pull request merged**